### PR TITLE
Switch builtin strings to use string tables

### PR DESCRIPTION
**Pull request by chandlerc**
The Clang binary (and any binary linking Clang as a library), when built using PIE, ends up with a pretty shocking number of dynamic relocations to apply to the executable image: roughly 400k.

Each of these takes up binary space in the executable, and perhaps most interestingly takes start-up time to apply the relocations.

The largest pattern I identified were the strings used to describe target builtins. The addresses of these string literals were stored into huge arrays, each one requiring a dynamic relocation. The way to avoid this is to design the target builtins to use a single large table of strings and offsets within the table for the individual strings. This switches the builtin management to such a scheme.

This saves over 100k dynamic relocations by my measurement, an over 25% reduction. Just looking at byte size improvements, using the `bloaty` tool to compare a newly built `clang` binary to an old one:

```
    FILE SIZE        VM SIZE
 --------------  --------------
  +1.4%  +653Ki  +1.4%  +653Ki    .rodata
  +0.0%    +960  +0.0%    +960    .text
  +0.0%    +197  +0.0%    +197    .dynstr
  +0.0%    +184  +0.0%    +184    .eh_frame
  +0.0%     +96  +0.0%     +96    .dynsym
  +0.0%     +40  +0.0%     +40    .eh_frame_hdr
  +114%     +32  [ = ]       0    [Unmapped]
  +0.0%     +20  +0.0%     +20    .gnu.hash
  +0.0%      +8  +0.0%      +8    .gnu.version
  +0.9%      +7  +0.9%      +7    [LOAD #2 [R]]
  [ = ]       0 -75.4% -3.00Ki    .relro_padding
 -16.1%  -802Ki -16.1%  -802Ki    .data.rel.ro
 -27.3% -2.52Mi -27.3% -2.52Mi    .rela.dyn
  -1.6% -2.66Mi  -1.6% -2.66Mi    TOTAL
```

We get a 16% reduction in the `.data.rel.ro` section, and nearly 30% reduction in `.rela.dyn` where those reloctaions are stored.

This is also visible in my benchmarking of binary start-up overhead at least:

```
Benchmark 1: ./old_clang --version
  Time (mean ± σ):      17.6 ms ±   1.5 ms    [User: 4.1 ms, System: 13.3 ms]
  Range (min … max):    14.2 ms …  22.8 ms    162 runs

Benchmark 2: ./new_clang --version
  Time (mean ± σ):      15.5 ms ±   1.4 ms    [User: 3.6 ms, System: 11.8 ms]
  Range (min … max):    12.4 ms …  20.3 ms    216 runs

Summary
  './new_clang --version' ran
    1.13 ± 0.14 times faster than './old_clang --version'
```

We get about 2ms faster `--version` runs. While there is a lot of noise in binary execution time, this delta is pretty consistent, and represents over 10% improvement. This is particularly interesting to me because for very short source files, repeatedly starting the `clang` binary is actually the dominant cost. For example, `configure` scripts running against the `clang` compiler are slow in large part because of binary start up time, not the time to process the actual inputs to the compiler.

----

This PR implements the string tables using `constexpr` code and the existing macro system. I understand that the builtins are moving towards a TableGen model, and if complete that would provide more options for modeling this. Unfortunately, that migration isn't complete, and even the parts that are migrated still rely on the ability to break out of the TableGen model and directly expand an X-macro style `BUILTIN(...)` textually. I looked at trying to complete the move to TableGen, but it would both require the difficult migration of the remaining targets, and solving some tricky problems with how to move away from any macro-based expansion.

I was also able to find a reasonably clean and effective way of doing this with the existing macros and some `constexpr` code that I think is clean enough to be a pretty good intermediate state, and maybe give a good target for the eventual TableGen solution. I was also able to factor the macros into set of consistent patterns that avoids a significant regression in overall boilerplate.

There is one challenge with this approach: it requires the host compiler to support (very) long string literals, a bit over half a meg. =/ The current minimum MSVC version rejects these, but the very next patch release (16.8) removed that restriction. I'm going to send out a separate PR / RFC to raise the minimum version by one patch release, which I hope is acceptable as the current version was set years ago.

FWIW, there are a few more low-hanging fruit sources of excessive dynamic relocations, maybe as many as 50k to 100k more that I'll take a look at to see if I can identify easy fixes. Beyond that, it seems to get quite difficult. It might be worth adding some guidance to developer documentation to try to avoid creating global data structures that _repeatedly_ store pointers to other globals.

### CODE REVIEW

**Comment by chandlerc** [comment]
Discussion thread about the MSVC version change: https://discourse.llvm.org/t/rfc-raising-minimum-msvc-version-by-one-patch-release/83490

**Review by dwblaikie — APPROVED** [review]
This looks fine to me - though given the broad impact (to every target in clang) maybe @AaronBallman wants to weigh in.

If you wanted to front-load the things like `getRecord(ID).Attributes` -> `getAttributesString(ID)` those things could possibly be done in advance.

**Review by zygoloid — APPROVED** [review]
LGTM too. I do wonder if we can make tablegen generate the data directly in the desired format here (perhaps with some additional `.td` files to define exactly which files we want to include in which targets and to define the placeholder for the target-independent builtins), but I don't think this (huge) improvement needs to block on that.

**Comment by chandlerc** [comment]
Updated to use the GCC diagnostic push the same as TableGen does for long string literals. Also added Aaron to take a look as well (unless he's comfortable already).

File: clang/lib/Basic/Targets/ARC.h
Comment (topperc) [inline clang/lib/Basic/Targets/ARC.h:43]: Should this override of getTargetBuiltins be removed? It looks like it matches the default.

Comment (chandlerc) [reply clang/lib/Basic/Targets/ARC.h]: Doh, yes. I meant to remove the base class one once they were all gone to make sure, and just forgot. Thanks for catching. Fixed.

File: clang/include/clang/Basic/Builtins.h
Comment (tbaederr) [inline clang/include/clang/Basic/Builtins.h:109]: This applies to a few functions added in this PR, but we don't really use trailing return types unless it's necessary

**Review by nikic — CHANGES_REQUESTED** [review]
Fails to build with GCC:
```

FAILED: tools/clang/lib/Basic/CMakeFiles/obj.clangBasic.dir/Builtins.cpp.o 
CCACHE_CPP2=yes CCACHE_HASHDIR=yes /usr/bin/ccache /usr/bin/c++ -DCLANG_EXPORTS -DGTEST_HAS_RTTI=0 -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/tmp/llvm-project-build-stage1/tools/clang/lib/Basic -I/var/llvm-compile-time-tracker/llvm-project/clang/lib/Basic -I/var/llvm-compile-time-tracker/llvm-project/clang/include -I/tmp/llvm-project-build-stage1/tools/clang/include -I/tmp/llvm-project-build-stage1/include -I/var/llvm-compile-time-tracker/llvm-project/llvm/include -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wno-misleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -fno-common -Woverloaded-virtual -fno-strict-aliasing -O3 -DNDEBUG  -fno-exceptions -funwind-tables -fno-rtti -std=c++17 -MD -MT tools/clang/lib/Basic/CMakeFiles/obj.clangBasic.dir/Builtins.cpp.o -MF tools/clang/lib/Basic/CMakeFiles/obj.clangBasic.dir/Builtins.cpp.o.d -o tools/clang/lib/Basic/CMakeFiles/obj.clangBasic.dir/Builtins.cpp.o -c /var/llvm-compile-time-tracker/llvm-project/clang/lib/Basic/Builtins.cpp
In file included from /var/llvm-compile-time-tracker/llvm-project/clang/lib/Basic/Builtins.cpp:13:
/var/llvm-compile-time-tracker/llvm-project/clang/include/clang/Basic/Builtins.h:137:3: error: ‘#pragma’ is not allowed here
  137 |   _Pragma("GCC diagnostic push")                                               \
      |   ^~~~~~~
/var/llvm-compile-time-tracker/llvm-project/clang/include/clang/Basic/Builtins.h:149:3: note: in expansion of macro ‘CLANG_BUILTIN_DETAIL_STR_TABLE’
  149 |   CLANG_BUILTIN_DETAIL_STR_TABLE(#ID "\0" TYPE "\0" ATTRS "\0" /*FEATURE*/ "\0")
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/llvm-compile-time-tracker/llvm-project/clang/lib/Basic/Builtins.cpp:34:9: note: in expansion of macro ‘CLANG_BUILTIN_STR_TABLE’
   34 |         CLANG_BUILTIN_STR_TABLE("not a builtin function", "", "")
      |         ^~~~~~~~~~~~~~~~~~~~~~~
/var/llvm-compile-time-tracker/llvm-project/clang/lib/Basic/Builtins.cpp: In member function ‘std::pair<const char*, const clang::Builtin::Info&> clang::Builtin::Context::getStrTableAndInfo(unsigned int) const’:
/var/llvm-compile-time-tracker/llvm-project/clang/lib/Basic/Builtins.cpp:48:65: error: could not convert ‘{<expression error>, <expression error>}’ from ‘<brace-enclosed initializer list>’ to ‘std::pair<const char*, const clang::Builtin::Info&>’
   48 |     return {BuiltinStorage.StringTable, BuiltinStorage.Infos[ID]};
      |                                                                 ^
      |                                                                 |
      |                                                                 <brace-enclosed initializer list>
```

File: clang/include/clang/Basic/Builtins.h
Comment (chandlerc) [reply clang/include/clang/Basic/Builtins.h]: Sorry, a new habit of mine and I hadn't gone through thoroughly enough to fix everywhere.

Should be fixed now throughout, I checked all the `auto` keywords in the diff.

**Review by chandlerc — COMMENTED** [review]
> Fails to build with GCC:

Doh, I had hoped GCC would support this. Anyways, fixed back to just suppress the diagnostic with Clang. Did some builds with GCC and everything seems fine, and it doesn't seem like we're GCC warning clean these days anyways.

File: clang/include/clang/Basic/Builtins.h
Comment (nikic) [inline clang/include/clang/Basic/Builtins.h:171]:
```suggestion
    sizeof(#ID), sizeof(TYPE), sizeof(ATTRS), sizeof("")                       \
```

File: clang/include/clang/Basic/Builtins.h
Comment (nikic) [inline clang/include/clang/Basic/Builtins.h:180]:
```suggestion
    sizeof(#ID), sizeof(TYPE), sizeof(ATTRS), sizeof(FEATURE)                  \
```

**Review by nikic — APPROVED** [review]
Confirmed that this works on GCC now. I'd suggest to replace the use of StringLiteral::size with plain sizeof(). The build time overhead of going through StringLiteral here is substantial.

**Comment by chandlerc** [comment]
> Confirmed that this works on GCC now. I'd suggest to replace the use of StringLiteral::size with plain sizeof(). The build time overhead of going through StringLiteral here is substantial.

Sure. I was initially worried about the subtlety of this use of `sizeof`, but that was before everything got nicely factored into macros. And yeah, the compile time cost is unfortunate.

Will merge once updated (including this fix) and builds passing.

**Comment by androm3da** [comment]
cc @llvm/pr-subscribers-backend-hexagon and @iajbar

**Comment by dyung** [comment]
> > LLVM Buildbot has detected a new failure on builder `llvm-clang-x86_64-sie-win` running on `sie-win-worker` while building `clang` at step 7 "test-build-unified-tree-check-all".
> > Full details are available at: [lab.llvm.org/buildbot#/builders/46/builds/9169](https://lab.llvm.org/buildbot/#/builders/46/builds/9169)
> 
> This is a Windows build bot with a version of MSVC '19.28.29924.0' -- that's not even one of the numbers listed here: https://en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B#Internal_version_numbering
> 
> (edit because I can't count)
> 
> Looks like it might be a build after 16.9.19 which is listed as 19.28.29923. Not sure what to do debug this... @zmodem maybe has some idea?
> 
> It doesn't seem to be the issue I know about on older versions of MSVC as there are no errors on the long string literal. But somehow it seems to be miscompiling the string literals?

I manage this build bot, is there anything I can do to help you debug this problem? We use this build as our "blessed" build internally for building on Windows.

**Comment by chandlerc** [comment]
> > > LLVM Buildbot has detected a new failure on builder `llvm-clang-x86_64-sie-win` running on `sie-win-worker` while building `clang` at step 7 "test-build-unified-tree-check-all".
> > > Full details are available at: [lab.llvm.org/buildbot#/builders/46/builds/9169](https://lab.llvm.org/buildbot/#/builders/46/builds/9169)
> > 
> > 
> > This is a Windows build bot with a version of MSVC '19.28.29924.0' -- that's not even one of the numbers listed here: [en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B#Internal_version_numbering](https://en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B#Internal_version_numbering)
> > (edit because I can't count)
> > Looks like it might be a build after 16.9.19 which is listed as 19.28.29923. Not sure what to do debug this... @zmodem maybe has some idea?
> > It doesn't seem to be the issue I know about on older versions of MSVC as there are no errors on the long string literal. But somehow it seems to be miscompiling the string literals?
> 
> I manage this build bot, is there anything I can do to help you debug this problem? We use this build as our "blessed" build internally for building on Windows.

Some indication of what's causing this? Or maybe whether there are any other versions of MSVC available with different results?

I don't have a Windows machine to debug on at all, but the premerge windows build was successful, so it seems to work on some builders but not others. So far I don't have any other failures to look at to cross-compare...

**Comment by dyung** [comment]
> > > > LLVM Buildbot has detected a new failure on builder `llvm-clang-x86_64-sie-win` running on `sie-win-worker` while building `clang` at step 7 "test-build-unified-tree-check-all".
> > > > Full details are available at: [lab.llvm.org/buildbot#/builders/46/builds/9169](https://lab.llvm.org/buildbot/#/builders/46/builds/9169)
> > > 
> > > 
> > > This is a Windows build bot with a version of MSVC '19.28.29924.0' -- that's not even one of the numbers listed here: [en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B#Internal_version_numbering](https://en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B#Internal_version_numbering)
> > > (edit because I can't count)
> > > Looks like it might be a build after 16.9.19 which is listed as 19.28.29923. Not sure what to do debug this... @zmodem maybe has some idea?
> > > It doesn't seem to be the issue I know about on older versions of MSVC as there are no errors on the long string literal. But somehow it seems to be miscompiling the string literals?
> > 
> > 
> > I manage this build bot, is there anything I can do to help you debug this problem? We use this build as our "blessed" build internally for building on Windows.
> 
> Some indication of what's causing this? Or maybe whether there are any other versions of MSVC available with different results?
> 
> I don't have a Windows machine to debug on at all, but the premerge windows build was successful, so it seems to work on some builders but not others. So far I don't have any other failures to look at to cross-compare...

I'll pull the machine offline and try to take a look. There appear to be two types of problems, 1 is an assertion failure in the pch related tests, the other use of an "unknown" builtin.

**Comment by chandlerc** [comment]
> > > > > LLVM Buildbot has detected a new failure on builder `llvm-clang-x86_64-sie-win` running on `sie-win-worker` while building `clang` at step 7 "test-build-unified-tree-check-all".
> > > > > Full details are available at: [lab.llvm.org/buildbot#/builders/46/builds/9169](https://lab.llvm.org/buildbot/#/builders/46/builds/9169)
> > > > 
> > > > 
> > > > This is a Windows build bot with a version of MSVC '19.28.29924.0' -- that's not even one of the numbers listed here: [en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B#Internal_version_numbering](https://en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B#Internal_version_numbering)
> > > > (edit because I can't count)
> > > > Looks like it might be a build after 16.9.19 which is listed as 19.28.29923. Not sure what to do debug this... @zmodem maybe has some idea?
> > > > It doesn't seem to be the issue I know about on older versions of MSVC as there are no errors on the long string literal. But somehow it seems to be miscompiling the string literals?
> > > 
> > > 
> > > I manage this build bot, is there anything I can do to help you debug this problem? We use this build as our "blessed" build internally for building on Windows.
> > 
> > 
> > Some indication of what's causing this? Or maybe whether there are any other versions of MSVC available with different results?
> > I don't have a Windows machine to debug on at all, but the premerge windows build was successful, so it seems to work on some builders but not others. So far I don't have any other failures to look at to cross-compare...
> 
> I'll pull the machine offline and try to take a look. There appear to be two types of problems, 1 is an assertion failure in the pch related tests, the other use of an "unknown" builtin.

I'm guessing they're the same issue, just a difference of whether hitting assert or a diagnostic first. Some strings for builtin names aren't coming back out of the tables correctly -- we fail to recognize it in one place, and hit an assert in the other.

My best guess is something going wrong with the string literals generated by the macros in this PR, or something going wrong with `sizeof` computing offsets... I'll do some experiments with compiler explorer, and see if I can get lucky...

If we need to revert the PR, totally understand, just a bit odd given only one compiler seems to be hitting the issue here.

**Comment by dyung** [comment]
> > > > > > LLVM Buildbot has detected a new failure on builder `llvm-clang-x86_64-sie-win` running on `sie-win-worker` while building `clang` at step 7 "test-build-unified-tree-check-all".
> > > > > > Full details are available at: [lab.llvm.org/buildbot#/builders/46/builds/9169](https://lab.llvm.org/buildbot/#/builders/46/builds/9169)
> > > > > 
> > > > > 
> > > > > This is a Windows build bot with a version of MSVC '19.28.29924.0' -- that's not even one of the numbers listed here: [en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B#Internal_version_numbering](https://en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B#Internal_version_numbering)
> > > > > (edit because I can't count)
> > > > > Looks like it might be a build after 16.9.19 which is listed as 19.28.29923. Not sure what to do debug this... @zmodem maybe has some idea?
> > > > > It doesn't seem to be the issue I know about on older versions of MSVC as there are no errors on the long string literal. But somehow it seems to be miscompiling the string literals?
> > > > 
> > > > 
> > > > I manage this build bot, is there anything I can do to help you debug this problem? We use this build as our "blessed" build internally for building on Windows.
> > > 
> > > 
> > > Some indication of what's causing this? Or maybe whether there are any other versions of MSVC available with different results?
> > > I don't have a Windows machine to debug on at all, but the premerge windows build was successful, so it seems to work on some builders but not others. So far I don't have any other failures to look at to cross-compare...
> > 
> > 
> > I'll pull the machine offline and try to take a look. There appear to be two types of problems, 1 is an assertion failure in the pch related tests, the other use of an "unknown" builtin.
> 
> I'm guessing they're the same issue, just a difference of whether hitting assert or a diagnostic first. Some strings for builtin names aren't coming back out of the tables correctly -- we fail to recognize it in one place, and hit an assert in the other.
> 
> My best guess is something going wrong with the string literals generated by the macros in this PR, or something going wrong with `sizeof` computing offsets... I'll do some experiments with compiler explorer, and see if I can get lucky...
> 
> If we need to revert the PR, totally understand, just a bit odd given only one compiler seems to be hitting the issue here.

Could the version of VC that we are using possibly have the issue with long strings that you mention? Is there a simple way to check that?

**Comment by chandlerc** [comment]
> Could the version of VC that we are using possibly have the issue with long strings that you mention? Is there a simple way to check that?

It's a compile time error, so no, that'd be really clear cut.

The only other thing I've seen is running out of heap, but that seems likely exclusively a compiler explorer thing.

**Comment by dyung** [comment]
I’m still looking into the failure, but may be a bit slow as the machine isn’t really setup for debugging, and I’m traveling today.

**Comment by zmodem** [comment]
> Not sure what to do debug this... @zmodem maybe has some idea?

Sorry, I don't have a lot to add. Things look good on my end so far (local builds and https://lab.llvm.org/buildbot/#/builders/63 + https://lab.llvm.org/buildbot/#/builders/107).

Besides the particular MSVC version on the failing bot, it's also using ccache. Perhaps that could be related somehow?

@dyung would it be possible to extract `tools\clang\lib\Basic\CMakeFiles\obj.clangBasic.dir\Builtins.cpp.obj` from one of the failing builds?

**Comment by chandlerc** [comment]
(FYI, I'm about to drop offline -- Hans, if you think it makes sense to revert temporarily, please feel free to do so. I don't have a good sense of whether the fallout here is enough to warrant that.)

**Comment by uweigand** [comment]
The s390x failure is just an unstable test that occasionally fails - that woudn't be a reason to revert.  Cannot say about the arm64-windows failure.

**Comment by chandlerc** [comment]
The arm64-windows failures are from `DirectoryWatcherTest` that seems exceedingly unlikely to be related.

I think the only real failures here are the originally discussed ones on the one windows build bot. All the other windows bots seems to be OK here.

**Comment by dyung** [comment]
> > Not sure what to do debug this... @zmodem maybe has some idea?
> 
> Sorry, I don't have a lot to add. Things look good on my end so far (local builds and https://lab.llvm.org/buildbot/#/builders/63 + https://lab.llvm.org/buildbot/#/builders/107).
> 
> Besides the particular MSVC version on the failing bot, it's also using ccache. Perhaps that could be related somehow?
> 
> @dyung would it be possible to extract `tools\clang\lib\Basic\CMakeFiles\obj.clangBasic.dir\Builtins.cpp.obj` from one of the failing builds?

Hmm... maybe it is ccache. Let me try doing a build with an empty cache and see if it still fails, otherwise, I'll extract that file.

**Comment by dyung** [comment]
> > > Not sure what to do debug this... @zmodem maybe has some idea?
> > 
> > 
> > Sorry, I don't have a lot to add. Things look good on my end so far (local builds and https://lab.llvm.org/buildbot/#/builders/63 + https://lab.llvm.org/buildbot/#/builders/107).
> > Besides the particular MSVC version on the failing bot, it's also using ccache. Perhaps that could be related somehow?
> > @dyung would it be possible to extract `tools\clang\lib\Basic\CMakeFiles\obj.clangBasic.dir\Builtins.cpp.obj` from one of the failing builds?
> 
> Hmm... maybe it is ccache. Let me try doing a build with an empty cache and see if it still fails, otherwise, I'll extract that file.

A clean build with an empty ccache does not seem to have made a difference. :(

I've pulled out the entire `tools\clang\lib\Basic` directory from a failing build, what would you like me to do with it?

**Comment by dyung** [comment]
> > I've pulled out the entire tools\clang\lib\Basic directory from a failing build, what would you like me to do with it?
> 
> Can you put it in a zip somewhere? It would be interesting to check if there's some major difference compared to using a different MSVC version.

I've put it up on Dropbox, let me know if you cannot access the zipfile: https://www.dropbox.com/scl/fi/pbc02zyba2wtcwhun27fz/Basic.zip?rlkey=v4vqd6bqoewubmkzca8508wxo&dl=0

**Comment by zmodem** [comment]
> I've put it up on Dropbox

I looked at Builtins.cpp.obj and compared to the one in my build dir. They're very similar, in particular the string table is there in both files, and looks identical. I don't really have any more ideas for debugging this.

**Comment by dyung** [comment]
No real updates here, but our internal builder did catch up to this commit and we are seeing the same (and a lot more) failures when this commit is merged into our downstream codebase. I was kind of hoping that it would pass so that it might indicate that the problem might be a configuration issue with our upstream build bot, but the fact that it fails in a similar manner likely points to some kind of bug in the version of MSVC that we are using.

@chandlerc could we consider reverting the change until we can figure out what the problem is? I will continue to take a look into it, but I don't really have any ideas what might be the cause at the moment, so have no ETA on a work-around, and in the meantime, this will continue to cause test failures for us internally.

**Comment by chandlerc** [comment]
> No real updates here, but our internal builder did catch up to this commit and we are seeing the same (and a lot more) failures when this commit is merged into our downstream codebase. I was kind of hoping that it would pass so that it might indicate that the problem might be a configuration issue with our upstream build bot, but the fact that it fails in a similar manner likely points to some kind of bug in the version of MSVC that we are using.

It really would be helpful to try a different version of MSVC.

As I pointed out in the original reply to the build bot failure, this bot seems to be using a version of MSVC that isn't even a listed release. The last patch release of 16.9 is `192829923` according to Wikipedia, but this builder reports `192829924`. To me, this seems to indicate a build that may not have gone through full Q&A and such.

Would it help for me to post a PR that adds some validation of the string table? I can easily do that, and maybe that would let you isolate whether this is definitely an MSVC version issue vs. something with the code.

> @chandlerc could we consider reverting the change until we can figure out what the problem is? I will continue to take a look into it, but I don't really have any ideas what might be the cause at the moment, so have no ETA on a work-around, and in the meantime, this will continue to cause test failures for us internally.

I'm happy to revert, and I'll start working on that. But I think there needs to be a concrete ETA on being able to make progress. Especially given the fact that no other builders seem to hit this and it may be specific to an MSVC version that even other windows devs don't realistically have access to...

Can you give some fairly concrete timeline? 1 week? maybe 2? Otherwise, there is no actual path to resolving this.

**Comment by dyung** [comment]
> > No real updates here, but our internal builder did catch up to this commit and we are seeing the same (and a lot more) failures when this commit is merged into our downstream codebase. I was kind of hoping that it would pass so that it might indicate that the problem might be a configuration issue with our upstream build bot, but the fact that it fails in a similar manner likely points to some kind of bug in the version of MSVC that we are using.
> 
> It really would be helpful to try a different version of MSVC.
> 
> As I pointed out in the original reply to the build bot failure, this bot seems to be using a version of MSVC that isn't even a listed release. The last patch release of 16.9 is `192829923` according to Wikipedia, but this builder reports `192829924`. To me, this seems to indicate a build that may not have gone through full Q&A and such.

The build was released as far as I'm aware (we don't install pre-release versions on our production machines). That being said, I'm guessing it is probably a bug with the specific version of MSVC that we are using. I have started the conversation internally about upgrading the version of MSVC we use to build, but that will take time.

> Would it help for me to post a PR that adds some validation of the string table? I can easily do that, and maybe that would let you isolate whether this is definitely an MSVC version issue vs. something with the code.

Sure, I'm happy to try anything that you think might help and this is one of my higher priority issues at the moment.

> > @chandlerc could we consider reverting the change until we can figure out what the problem is? I will continue to take a look into it, but I don't really have any ideas what might be the cause at the moment, so have no ETA on a work-around, and in the meantime, this will continue to cause test failures for us internally.
> 
> I'm happy to revert, and I'll start working on that. But I think there needs to be a concrete ETA on being able to make progress. Especially given the fact that no other builders seem to hit this and it may be specific to an MSVC version that even other windows devs don't realistically have access to...
> 
> Can you give some fairly concrete timeline? 1 week? maybe 2? Otherwise, there is no actual path to resolving this.

I will look into it the rest of this week and into next week. I don't really know the problem here, so I cannot promise a resolution date, but I will try my best to get this resolved, it is making our internal automation fail, so I want to unblock that!

**Comment by dyung** [comment]
> So, the builder came back up, and is now passing without this landing...
> 
> Here is the first build after it came back up: https://lab.llvm.org/buildbot/#/builders/46/builds/9173
> 
> And I went through the output and found a specific test that was failing, and it seems to pass?
> 
> ```
> PASS: Clang :: AST/builtins-arm-strex-rettype.c (433 of 81898)
> ```
> 
> It looks like the builder is now using MSVC version 14.29.30133 -- so that might explain it. This maybe means that there is some issue with later patch releases of 14.28 (in Visual Studio v16.9 as I understand it)?
> 
> Are all of your builders OK moving to 14.29 (Visual Studio 16.11 from my reading)?

That was a mistake. I originally was going to update the build bot to a later version of Visual Studio as while it wouldn't match our internal configuration, it was still useful to have some Windows coverage of the components we build/test. I then changed my mind and was just going to restore the original configuration (and thought I had) but had accidentally replaced the compiler with a newer version. I'm going to put the original one back.

I have started the process to try and upgrade the version we are using internally, but that will likely take a while.

**Comment by chandlerc** [comment]
> That was a mistake. I originally was going to update the build bot to a later version of Visual Studio as while it wouldn't match our internal configuration, it was still useful to have some Windows coverage of the components we build/test. I then changed my mind and was just going to restore the original configuration (and thought I had) but had accidentally replaced the compiler with a newer version. I'm going to put the original one back.

Ah, no problem.

Let me know when its up and I can wait for at least one build to cycle before reverting.

> 
> I have started the process to try and upgrade the version we are using internally, but that will likely take a while.

Of course, that's part of why I asked -- I was quite surprised to see the update.

This does at least seem to isolate to MSVC version though and not some other aspect of the machine or configuration, which is actually really useful to know.

**Comment by chandlerc** [comment]
I see the build bot up and with the original failure, going to merge the revert to hopefully make sure it goes green after that.

**Pull request merged**
